src/python/zquantum/core/wip/circuit/conversions2/cirq_conversions.py

```diff
@@ -1 +1,242 @@
+
+
+def make_rotation_factory(
+    eigengate_cls: Type[cirq.EigenGate], global_shift: float = 0
+) -> RotationGateFactory:
+    """Construct a factory for rotation gate based on given EigenGate subclass.
+
+    This function might be thought of as a partial which freezes global_shift
+    parameter but also scales the exponent parameter of eigengate_cls initializer.
+
+    Args:
+        eigengate_cls: EigenGate subclass, e.g. ZPowGate, XXPowGate.
+        global_shift: Determines phase of the rotation gate. Check Cirq docs
+            for explanation.
+    Returns:
+        A function that maps angle to EigenGate instance with given global shift
+        and an exponent equal to angle divided by a factor of pi.
+    """
+
+    def _rotation(angle: Parameter) -> cirq.EigenGate:
+        return eigengate_cls(
+            global_shift=global_shift, exponent=angle_to_exponent(angle)
+        )
+
+    return _rotation
+
+
+ORQUESTRA_BUILTIN_GATE_NAME_TO_CIRQ_GATE = {
+    "X": cirq.X,
+    "Y": cirq.Y,
+    "Z": cirq.Z,
+    "I": cirq.I,
+    "H": cirq.H,
+    "T": cirq.T,
+    "RX": cirq.rx,
+    "RY": cirq.ry,
+    "RZ": cirq.rz,
+    "PHASE": make_rotation_factory(cirq.ZPowGate),
+    "CNOT": cirq.CNOT,
+    "CZ": cirq.CZ,
+    "SWAP": cirq.SWAP,
+    "ISWAP": cirq.ISWAP,
+    "CPHASE": cirq.cphase,
+    "XX": make_rotation_factory(cirq.XXPowGate, -0.5),
+    "YY": make_rotation_factory(cirq.YYPowGate, -0.5),
+    "ZZ": make_rotation_factory(cirq.ZZPowGate, -0.5),
+    "XY": make_rotation_factory(cirq.ISwapPowGate, 0.0),
+}
+
+
+EIGENGATE_SPECIAL_CASES = {
+    (type(cirq.X), cirq.X.global_shift, cirq.X.exponent): _builtin_gates.X,
+    (type(cirq.Y), cirq.Y.global_shift, cirq.Y.exponent): _builtin_gates.Y,
+    (type(cirq.Z), cirq.Z.global_shift, cirq.Z.exponent): _builtin_gates.Z,
+    (type(cirq.T), cirq.T.global_shift, cirq.T.exponent): _builtin_gates.T,
+    (type(cirq.H), cirq.H.global_shift, cirq.H.exponent): _builtin_gates.H,
+    (type(cirq.CNOT), cirq.CNOT.global_shift, cirq.CNOT.exponent): _builtin_gates.CNOT,
+    (type(cirq.CZ), cirq.CZ.global_shift, cirq.CZ.exponent): _builtin_gates.CZ,
+    (type(cirq.SWAP), cirq.SWAP.global_shift, cirq.SWAP.exponent): _builtin_gates.SWAP,
+    (type(cirq.ISWAP), cirq.ISWAP.global_shift, cirq.ISWAP.exponent): _builtin_gates.ISWAP,
+}
+
+EIGENGATE_ROTATIONS = {
+    (cirq.XPowGate, -0.5): _builtin_gates.RX,
+    (cirq.YPowGate, -0.5): _builtin_gates.RY,
+    (cirq.ZPowGate, -0.5): _builtin_gates.RZ,
+    (cirq.ZPowGate, 0): _builtin_gates.PHASE,
+    (cirq.CZPowGate, 0): _builtin_gates.CPHASE,
+    (cirq.XXPowGate, -0.5): _builtin_gates.XX,
+    (cirq.YYPowGate, -0.5): _builtin_gates.YY,
+    (cirq.ZZPowGate, -0.5): _builtin_gates.ZZ,
+    (cirq.ISwapPowGate, 0.0): _builtin_gates.XY,
+}
+
+CIRQ_GATE_SPECIAL_CASES = {cirq.CSWAP: _builtin_gates.SWAP.controlled(1)}
+
+qubit_index = attrgetter("x")
+
+
+@singledispatch
+def export_to_cirq(obj):
```

Comment:
WDYT about providing a type hint for `obj` here? Without it, the user has to read through `singledispatch` decorators to see what objects can be exported.